src/failure_message.rs

```diff
@@ -0,0 +1,97 @@
+use lazy_static::lazy_static;
+use phf::phf_ordered_map;
+
+use pyo3::{ffi::PySys_ResetWarnOptions, prelude::*, types::PyString};
+
+use itertools::Itertools;
+use regex::Regex;
+
+// Need to use an ordered map to make sure we replace '>' before
+// we replace '\n', so that we don't replace the '>' in '<br>'
+static REPLACEMENTS: phf::OrderedMap<&'static str, &'static str> = phf_ordered_map! {
+    "\"" => "&quot;",
+    "'" => "&apos;",
+    "<" => "&lt;",
+    ">" => "&gt;",
+    "&" => "&amp;",
+    "\r" =>  "",
+    "\n" =>  "<br>",
+};
+
+#[pyfunction]
+pub fn escape_failure_message_pystring<'py>(
+    py: Python<'py>,
+    failure_message: &PyString,
+) -> &'py PyString {
+    let mut escaped_failure_message = failure_message.to_string();
+    for (from, to) in REPLACEMENTS.entries() {
+        escaped_failure_message = escaped_failure_message.replace(from, to);
+    }
+    &PyString::new(py, &escaped_failure_message)
+}
+
+#[pyfunction]
```

Comment:
If I'm in python code is it possible to generate a rust string to pass to this function?

Also I didn't really understand why we need both of them exposed to the python code...